vendor/zkllvm-metacraft-circuits/src/circuit_utils/base_types.h

```diff
@@ -21,29 +28,62 @@
     { assert(c); }
 #endif
 
-using Byte = unsigned char;
-using Bytes32 = std::array<Byte, 32>;
-using Bytes64 = std::array<Byte, 64>;
+using sha256_t = typename nil::crypto3::hashes::sha2<256>::block_type;
+
+#ifdef __ZKLLVM__
+using HashType = sha256_t;
+#else
+using HashType = Bytes32;
+#endif
 
 using Epoch = uint64_t;
 using Slot = uint64_t;
-using Root = Bytes32;
+using Root = HashType;
 using Gwei = uint64_t;
 template<size_t DEPTH>
-using MerkleProof = std::array<Bytes32, DEPTH>;
+using MerkleProof = static_vector<HashType, DEPTH, true>;
 using BeaconStateLeafProof = MerkleProof<5>;
 
+#define countof(array) (sizeof(array) / sizeof(array[0]))
+
+bool sha256_equals(sha256_t hash1, sha256_t hash2) {
+    bool result = true;
+    for (auto i = 0; i < countof(hash1); ++i) {
+        result = result && (hash1[i] == hash2[i]);
+    }
+
+    return result;
+}
+
+bool sha256_equals(Bytes32 hash1, Bytes32 hash2) {
+    return hash1 == hash2;
+}
+
 struct CheckpointVariable {
     Epoch epoch;
     Root root;
     bool operator==(const CheckpointVariable &c) const {
-        return (epoch == c.epoch && root == c.root);
+        return (epoch == c.epoch && sha256_equals(root, c.root));
     }
-};
+} __attribute__((packed));
 
 struct JustificationBitsVariable {
 
-    std::array<bool, 4> bits;
+    static_vector<bool, 4, true> bits;
+
+    constexpr JustificationBitsVariable(std::initializer_list<bool> init) {
+        size_t i = 0;
+        for (const auto &v : init) {
+            assert_true(i < bits.size());
```

Comment:
Use initializer_list's constexpr size to do the check once

Reply:
You could static_assert that the size is 4

Reply:
I have changed the check to be performed once. Making it a static_assert, though causes a problem
```
/zkllvm-template/src/circuit_utils/base_types.h:75:35: error: non-constant condition for static assertion
   75 |         static_assert(init.size() == bits.capacity());
      |                       ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
/zkllvm-template/src/circuit_utils/base_types.h:75:32:   in 'constexpr' expansion of 'init.std::initializer_list<bool>::size()'
/zkllvm-template/src/circuit_utils/base_types.h:75:35: error: 'init' is not a constant expression
```

Reply:
ah yes, you could pass an std::array of size 4

Reply:
Since it is better to have compile time check, it is worth having different initialization call. Will change it.